### ISSUE TITLE
Fix description of `IN_UNIT`

### ISSUE THREAD

**Issue by gmantele** [ivoa-std/ADQL#18]
`IN_UNIT` is a function able to convert a value into another unit. Its current signature is:

```
IN_UNIT(value, target_unit)
```

This works as expected if the 1st argument is a reference to a column whose the unit is specified in `TAP_SCHEMA.columns`. But what happens ...

- ... if no unit is set for this column reference?
- ... if the value is not a column reference (e.g. a numeric litteral, a math expression)?_

Nothing is specified in the current state of the document. Hence this issue.

During the IVOA Interop. Meeting (taking place in Groningen in Oct. 2019), there was a consensus for the following solution:

> If no unit metadata is available, `IN_UNIT` should return the given value unchanged with the unit piece of metadata set to the given target unit. In other words, in such case: no conversion, just annotation.

**Comment by gervandiepen** [comment]
The IN_UNIT documentation in WD-ADQL-2.1-20171208.pdf is quite sparse.
I have several questions/remarks.

What happens if an operation (such as addition, comparison) is done on 2 values having different but compatible units. Will the second value automatically be converted to the first unit?
Will the result of an addition of 2 values with the same unit have that unit?
Similarly, what happens if e.g. the cosine is taken of a value with unit DEG?
Can the value of IN_UNIT be a set, so you can convert a set of values?
Are prefixes (such as k in km or kparsec) supported?
Are conversions of compound units supported (such as m/s to km/h)?

I guess the above is not supported, so it should be stated more explicitly in the section describing IN_UNIT that the unit attribute is only used by IN_UNIT and is lost if any operation is done on values. Maybe also state that the data type of the IN_UNIT value argument must be numeric (thus no string, datetime, etc.).

BTW. I would very much like the following (as possible in casacore's TaQL):
     2 deg rad     or if really needed:  2 "deg" "rad"
instead of
     IN_UNIT(IN_UNIT(2, "deg"), "rad")
Not only is it less verbose, it is also more clear.
Note that in the first line the unit can be seen as a postfix operator. It can be parsed unambigiously because in SQL (and other languages) a value or subexpression can never be followed by a string (except for operators such as IN which are reserved words).
However, I realize this is something that cannot be adopted easily in ADQL.

**Comment by msdemlei** [comment]
Hi Ger,

On Tue, Jan 21, 2020 at 03:23:29AM -0800, Ger van Diepen wrote:
> What happens if an operation (such as addition, comparison) is done
> on 2 values having different but compatible units. Will the second
> value automatically be converted to the first unit?

No.  ADQL as such has no notion of units.  IN_UNIT is a minimal
attempt to remedy this, but only at the level of the translation
engine.  The result of IN_UNIT is a normal SQL/ADQL value, and no
magic will happen when it appears in expressions.

Unit inference happens at the level of the translation engine and,
for the purpose of ADQL, is an implementation detail.

> Will the result of an addition of 2 values with the same unit have
> that unit?

The rules to infer units (and, for that matter, UCDs) are left to the
engines as implemenation details.  It would certainly be nice if we
gave a strict calculus for them, but both VOUnit and ADQL's operators
are so complex that that would be quite a bit of material.  And given
that you currently deal with literals ("3600") in many places that
really carry units ("3600 s/h"), it would still break regularly.

> Similarly, what happens if e.g. the cosine is taken of a value with
> unit DEG?

You'll get a wrong result, as before.  IN_UNIT doesn't make ADQL
unit-aware.

Concrete implementations might warn against such things, though, with
or without in_unit [which brings me back to my old with for a good
way to convey such warnings to users].

> Can the value of IN_UNIT be a set, so you can convert a set of values?

That is an excellent question, as it has made me realise the grammar
reserves IN_UNIT but then fails to mention it elsewhere in the
grammar.  Ouch.  I'll file a bug.

Anyway, I'd strongly suggest we have 

in_unit ::= "IN_UNIT" "(" numeric_value_expression ","
  character_string_literal ")"

-- and hence, no, set-valued expressions are not allowed.  

Relatedly, ADQL doesn't yet talk much about arrays, but
numeric_value_expression can be a column reference, and hence
implementations having these in their tables might want to worry
about IN_UNIT and arrays.  Disclosure: DaCHS doesn't worry so far, and
users will receive interesting error messages when they apply IN_UNIT
to arrays.

> Are prefixes (such as k in km or kparsec) supported?

Yes -- the unit syntax is defined by VOUnit and allows (and rather
rigorously defines) prefixes (and much beyond), including declaring
that furlong should parse as femto-urlong (which always has been my
favourite part of the spec).

> Are conversions of compound units supported (such as m/s to km/h)?

Yes, that is implied by the reference to VOUnits.  To what extent an
implementation does this is another question.  For instance, DaCHS
will not successfully convert log(solMass) -- a legal VOUnit unit --
to kg (though that might be reasonable).  And I've long thought about
whether anything sensible can be done with IN_UNIT(freq, "Angstrom")
with freq in GHz (and decided no).

> I guess the above is not supported, so it should be stated more
> explicitly in the section describing IN_UNIT that the unit
> attribute is only used by IN_UNIT and is lost if any operation is
> done on values. Maybe also state that the data type of the IN_UNIT
> value argument must be numeric (thus no string, datetime, etc.).

On making clear that the return value of IN_UNIT is not somehow
"blessed": Yes, I think that would be an improvement.  If you have a
concrete proposal for a text, I'm sure the editor would be grateful.

On restricting the argument type: Beyond the syntactic constraints,
I'd be hesitant to state further restrictions; I do not think they
add much over the "The system should report an error [... if it] is
not able to convert the value into the required units" -- that
already catches timestamps and strings, which are unitless.  And
additional constraints may introduce unwelcome corner cases.

> BTW. I would very much like the following (as possible in casacore's TaQL):
>      2 deg rad     or if really needed:  2 "deg" "rad"
> instead of
>      IN_UNIT(IN_UNIT(2, "deg"), "rad")
> Not only is it less verbose, it is also more clear.

For the record: You would say "IN_UNIT(col, 'rad')" and rely on the
database to work out the unit of col; giving literals units in this
way is not intended and should fail: IN_UNIT(2, 'deg') is a request
to convert a dimensionless 2 to degrees, which can't work.

As to your basic statement: Yes, the creators of ADQL could, and
perhaps should, have put in a unit calculus.  If I did it now, I'd
probably follow the astropy example, complete with all the magic they
do there.

From where we are now, however, as you say:

> However, I realize this is something that cannot be adopted easily
> in ADQL.

-- I suppose adding implicit unit handling would hurt in many
places (including quite possibly changing results of queries).  So, I
don't think there is a way we can evolve ADQL 2 into that direction.

        -- Markus

**Comment by gervandiepen** [comment]
Hi Markus,

Thanks for your answer. I expected about all of it, but wanted it to be more clear.
Regarding assigning a unit to a constant. From the issue description

> During the IVOA Interop. Meeting (taking place in Groningen in Oct. 2019), there was a consensus for the following solution:
> 
>     If no unit metadata is available, IN_UNIT should return the given value unchanged with the unit piece of metadata set to the given target unit. In other words, in such case: no conversion, just annotation.

Furthermore, in your answer you propose to define in_unit as

in_unit ::= "IN_UNIT" "(" numeric_value_expression ","
  character_string_literal ")"

From both I conclude that it should be possible to attach a unit to a numeric constant and to convert it as done in my example.
From your answer it seems that IN_UNIT can only be used to convert (or define) the unit of a column. If so, that should certainly be made more clear.

Units in astropy require Python functionality, whereas ADQL is a command-oriented interface. My (strongly biased :-) opinion is that TaQL is a good example how units could be integrated. Your
furlong example (f is handled differently):
TaQL> 1 fur m
Unit: m
201.168
TaQL> (42.2km/3h34m17 h)'m/s'    ;sexagesimal gives rad which is converted to h(our)
Unit: m/s
3.28226

BTW. Can sexagesimal format (1h23m45.67 or 1d23m45.67) be used in ADQL? I could not find it in the document.

**Comment by msdemlei** [comment]
On Tue, Jan 21, 2020 at 06:08:44AM -0800, Ger van Diepen wrote:
> > During the IVOA Interop. Meeting (taking place in Groningen in Oct. 2019), there was a consensus for the following solution:
> > 
> >     If no unit metadata is available, IN_UNIT should return the given value unchanged with the unit piece of metadata set to the given target unit. In other words, in such case: no conversion, just annotation.

Hm.  Now that I read it like this, I think I have to revoke the
consensus.  Here's my reasoning: At least my plan has been that
people can say something like

dec+IN_UNIT(pmdec, "deg/yr")*epdiff

(disregarding that that only works satisfactorily for small pmdec and
epdiff) and confidently run this on different services without having
to worry about the actual unit of pmdec.

If the service could effectively ignore IN_UNIT (rather than raising an
error), that confidence goes out of the door, as pmdec would just
fall thorugh and be added in whatever unit it really happens to be.

I'm fairly sure that we need to have: "Whenever the inferred unit of
value is incommensurable with target_unit (which includes the cases
that value has no unit or the unit cannot be inferred), the service
must return an error."


> Furthermore, in your answer you propose to define in_unit as
> 
> in_unit ::= "IN_UNIT" "(" numeric_value_expression ","
>   character_string_literal ")"
> 
> From both I conclude that it should be possible to attach a unit to
> a numeric constant and to convert it as done in my example.

Oh, the fact that you can syntactically pass in numeric literals
doesn't mean IN_UNIT can't raise an error on semantic grounds.

> BTW. Can sexagesimal format (1h23m45.67 or 1d23m45.67) be used in
> ADQL? I could not find it in the document.

It cannot -- ADQL floats are SQL floats, and they are solidly decimal
[anti-Babylonian rant elided].

**Comment by gmantele** [comment]
Let me try to sum up all these ideas.

I agree with the following points that ADQL-2.1 must specify:

1. the following piece of BNF must be added:

   ```
   <in_unit> ::= IN_UNIT <left_paren> <numeric_value_expression>
                              <comma> <character_string_literal> <right_paren>
   ```

   and `numeric_value_function` must be updated like this:
   
   ```
   <numeric_value_function> ::=
       <trig_function>
     | <math_function>
     | <in_unit>
     | <numeric_geometry_function >
     | <user_defined_function>
   ```

2. the 1st argument of `IN_UNIT` MUST be a numeric value.

3. the 2nd argument MUST be a string litteral whose format is defined by
   [VOUnits Specification](http://www.ivoa.net/documents/VOUnits/20140523/VOUnits-REC-1.0-20140523.pdf)

4. whether or not the unit of the result should be computed automatically by the ADQL
   implementation

----------

Then, I think a bit more discussion is needed on the following idea:

```
If no unit is specified, `IN_UNIT` sets the unit metadata of the resulting value
to the given unit.
```
   
@msdemlei seems to disagree on this point. Since I was the one proposing this
statement in first place, let me just precise my thoughts at that moment.

`IN_UNIT`, as any other ADQL function, can be used in the SELECT clause.
In such case, there will be a `FIELD` in the VOTable output. This field will
have metadata...which include the unit. That was a first reason why `IN_UNIT`
should be allowed to set the unit of a litteral value (or operation). Besides,
this can be particularly useful if the result of a such query is uploaded to an
ADQL-based service (e.g. TAP).

Then, allowing such thing would also let user write operations
(e.g. a multiplication) between a column having a unit provided by the
service/database (e.g. `m`) and a constant or another operation that the ADQL
user knows it has a unit as well (e.g. `m` again).  In such case, the result
would have to be in the correct unit (e.g. `m**2`). I admit that it is a case
that won't happen that often, but at least it is a possibility that few users
may like to have.

So, for a litteral or the result of an operation, I think it should be
allowed: there is no other way to specify the unit of a such value and
sometimes it can be useful (for `SELECT` output or operation as explained
above).

However, allowing `IN_UNIT` set a unit on unitless columns is a bit more
tricky as it would mean that the service provider did not do his job
correctly by omitting to provide (all or some) units. Anyway, I think it
should be allowed. First, because otherwise ADQL-2.1 would have to specify an
exception (which always complicate implementations and usage). And then, because
columns may come from the database but also from subqueries or TAP_UPLOADs where
metadata are more frequently incomplete.


----------

Finally, I like the idea of suffixing litteral values with their unit (as in
TaQL). I think it makes it more explicit for humans. But as you both said, it is
difficult to make that possible in the current state of ADQL. So that is
probably an idea to keep in mind for a future version.

**Comment by msdemlei** [comment]
On Wed, Jan 29, 2020 at 08:00:37AM -0800, Grégory Mantelet wrote:
> Then, I think a bit more discussion is needed on the following idea:
> 
> ```
> If no unit is specified, `IN_UNIT` sets the unit metadata of the resulting value
> to the given unit.
> ```
>    
> @msdemlei seems to disagree on this point. Since I was the one proposing this
> statement in first place, let me just precise my thoughts at that moment.
> 
> `IN_UNIT`, as any other ADQL function, can be used in the SELECT clause.
> In such case, there will be a `FIELD` in the VOTable output. This field will
> have metadata...which include the unit. That was a first reason why `IN_UNIT`
> should be allowed to set the unit of a litteral value (or operation). Besides,
> this can be particularly useful if the result of a such query is uploaded to an
> ADQL-based service (e.g. TAP).

While I agree that your use case -- force metadata when the engine
can't infer it or, worse, get it wrong -- is a valid one, and indeed
I'd even extend it to include forcing UCDs, I'd argue IN_UNIT is the
wrong tool to satisfy it.

My main intention with IN_UNIT was that as user, you can be sure the
result is in the unit you're giving (excluding data bugs; I expect
there'll be pmra=-99.0 for a long time yet).

To make this work, services must reliably flag errors when they can't
do the conversion for whatever reason.  They can't do that if they
have to allow the assignment of metadata.  Consider

  IN_UNIT(expr, 'deg/yr')

Let's assume we'd allow assigning units with IN_UNIT.  The service
would then have to reason somewhat like this:

  If the service can determine a unit u on expr

    if u is compatible with 'deg/yr', 
    
      compute the conversion function and generate the appropriate
      expression, and set expr's (internal) unit metadata to deg/yr.

    else

      raise an error about incompatible units

  else

    leave expr alone, just set the internal unit metadata.

While this might sound reasonable at first sight, the problem is the
top-level if: There are just too many reasons why the service might
not be able to determine "a" unit of expr, for instance:

* expr might be unitless to begin with; at least DaCHS isn't terribly
  concerned between a truly empty unit (e.g., [Fe/H], which is a
  ratio of things with units -- and no, dex is not a unit, it's minor
  scientific delinquency) and something that doesn't have units to
  begin with (e.g., n_obs; though trolls might argue it should really
  have count as unit).  In the first case, I guess we agree IN_UNIT
  out to error out.  The second case... hm.  I don't even know.  You
  don't want to have such considerations in your ADQL translator.

* col might be a complex expression that *should* have a unit, but
  the service has given up at some point (for instance, current DaCHS
  won't figure out the unit of log10(power(10, pmra))).  I'd argue at
  least there, the system must raise an error for reliability.

* col might be a literal.  I'd agree that if we used IN_UNIT for the
  use case, this should not error out and instead set unit metadata.

Having to pick apart these different cases is hard to specify and
brittle in implemenation.  Let's not do it.

> Then, allowing such thing would also let user write operations
> (e.g. a multiplication) between a column having a unit provided by
> the service/database (e.g. `m`) and a constant or another operation
> that the ADQL user knows it has a unit as well (e.g. `m` again).
> In such case, the result would have to be in the correct unit (e.g.
> `m**2`). I admit that it is a case that won't happen that often,
> but at least it is a possibility that few users may like to have.

I don't think that's a terribly rare thing to want.  It would be an
interesting exercise to look at astropy programmes (or, sigh,
notebooks) to see what people do with the unit calculus people there
have.  I just think that overloading IN_UNIT will spoil it for both
of these diffent use cases.

I think what we should look at is some function/syntax of injecting
metadata into ADQL queries.  The IMHO cleanest way would be a
function

  FORCE_META(expr, unit, [ucd])

as in 

  SELECT IN_UNIT(pmra, 'mas/h')*FORCE_META(3600, 's/h')...

-- the ADQL engine can then work out the the result is in mas/s,
which is nice.

However, I give you that's a mouthful to write.  Now, ADQL right now
doesn't do anything with ! -- perhaps that's what we ought to do:

  3600!'s/h'

I've thought about "@", but IMHO the bang suggests the forcefulness
of the operation, because I'm very sure something like
(pmra*3600)!'m/s' should just blindly put m/s as unit and ask no
questions (where IN_UNIT would howl and spit).

But all that would be for ADQL 2.2.  I don't think we can get
anything like this into ADQL 2.1, and as I said I'm sure IN_UNIT
isn't the right vehicle.

**Comment by gervandiepen** [comment]
It would be nice if it could be specified in the VOTable if columns have no unit, so an error can be raised if a unit is set for such a column. Having no unit is different from not having specified if a column has a unit.

I think units are very useful in trigonometric functions, regions, etc. Not having to do an explicit conversion and not having to know if the argument should be rad or deg is very nice.
It requires that ADQL has to know the unit of function arguments and results, also of possible user-defined functions. Maybe something for a future version.

As such I see no problem in using IN_UNIT to set a unit for a constant value or a column without unit definition, but if it has to be more explicit FORCE_META could be used. I hope it will not confuse users.

TaQL users regularly use units, for instance to get the result in the unit they prefer (such as MHz). 
        SELECT FREQ MHz from table
But also when using functions taking angles such as cos or cone search they do not have to think about the unit of the arguments.
Note that in TaQL every expression (also columns) can be suffixed by a unit which sets the unit or does unit conversion.

For a future version of ADQL I argue for sexagesimal constants which can easily be recognised and converted to a float constant (with a unit :-) by the ADQL scanner part. In my experience astronomers often think that way.

**Comment by gmantele** [comment]
As @gervandiepen , I also think that the problem is coming from the absence of
a way to specify that a value is unitless. If we have a such special value,
`IN_UNIT` could set a unit on a litteral/column/operation having no unit
specified, as well as raising an error in the case the set unit is `unitless`.

------

I am no expert in VOUnits, but is there really no way to say `unitless`?
Would it be something to add in there?

In the meantime, in VOTable, could we assume that an empty attribute value
would mean `unitless` and that if the attribute is missing no unit is specified?

------

Yet, I think that `IN_UNIT` should be allowed to set unit on a quantity whose
unit is unspecified. If one day VOUnits is updated and provide a `unitless`
option, the ADQL standard would not need to be updated because a conversion from
`unitless` to anything else will systematically raise an error.

I don't really like the idea of the function `FORCE_META`. Having a such
function makes ADQL able to modify the formatted output: ADQL is no longer
selecting records but lets modify its full attached metadata. That would be a
mix between selecting and formating in a single _query_ language...I don't think
it is a great idea to make ADQL that powerful ; metadata management should be
the responsibility of the service, not the language. Plus, it is then not clear
what `FORCE_META` should return: obviously you mean the argument, but the
reading of the ADQL query is less readable.

The proposition of `IN_UNIT` setting the unit metadata is essentially useful in
operations. Setting the unit metadata field when used in `SELECT` is only a very
convenient side-effect. In my mind the unit metadata setting ability of
`IN_UNIT` is not its main goal. Its main goal is to convert a value from one
unit to another, but ADQL has to say what to do if no unit is specified. Then,
we have 2 choices: 1/ raise an error, 2/ set the unit to the given one. Raising
an error seems a bit extrem for me and will probably disturb users since
probably a lot of columns (from DB or UPLOAD), litterals or operations may not
have unit. Then, I think it is reasonnable to assume that the user SHOULD know
what he is doing, scientifically speaking.

------

> For a future version of ADQL I argue for sexagesimal constants which can easily be recognised and converted to a float constant (with a unit :-) by the ADQL scanner part. In my experience astronomers often think that way.

I also agree that sexagesimal coordinates are way more frequently used by
astronomers than degrees. We should probably think to a way to allow that in a
future version of ADQL...probably a function like `sexa2deg` (sorry, I am not
really good in inventing names :-( ) or it would be something going along a
TaQL-like way to specify unit of constants.

**Comment by msdemlei** [comment]
On Thu, Jan 30, 2020 at 06:02:55AM -0800, Grégory Mantelet wrote:
> I am no expert in VOUnits, but is there really no way to say `unitless`?
> Would it be something to add in there?

The understanding in VOUnit has always been that an empty string
corresponds to unitless.

The trouble is more on the plumbing side, where service and client
software isn't really careful about distinguishing between "unitless"
and "thing that doesn't have a unit".  And that's partly because that's
hard, as by n_obs example shows -- some would argue that such a number
can't have a unit to begin with, others would happily put a "count"
on it.

> In the meantime, in VOTable, could we assume that an empty
> attribute value would mean `unitless` and that if the attribute is
> missing no unit is specified?

We could, and perhaps that would be something we should add on the
wishlist for VOTable 1.5.  But that doesn't save you from the fact
that that distinction isn't simple and isn't done very well out
there.

> Yet, I think that `IN_UNIT` should be allowed to set unit on a
> quantity whose unit is unspecified. If one day VOUnits is updated
> and provide a `unitless` option, the ADQL standard would not need
> to be updated because a conversion from `unitless` to anything else
> will systematically raise an error.

Indulge me for a second and let me try to explain again why I think
any such functionality would be a mistake.  Shout where you disagree:

(1) IN_UNIT(ra, "deg/yr"), where ra is in deg must error out, right?  If
we don't have that, it's far too easy to write nonsense; actually, I
think the primary use case of IN_UNIT will be by scripts that figure out
columns by UCD and want to write simple queries based on this; if
something goes wrong in UCD reasoning, that's another safeguard.  Also:
What should an engine do?

(2) IN_UNIT(fe_h, "solMass"), where fe_h is unitless, should error out,
right?  The only use case I could see for such a thing would be to fix
missing annotation, and tehn you're tailoring your query to the actual
service (since the next service will in all likelihood have different
annotation problems) -- so, why use IN_UNIT in the first place if you
know what to do?

(3) IN_UNIT(n_obs, "count"), where n_obs is the number of observations
that went into a row, should error out.  The reasoning goes as for (2).

(4) IN_UNIT(log10(power(v), 10), "km/s") with v in m/s might error out
if the service can't work out that the unit of the first argument is
exactly the unit of v.  If you drop that, it's again far too simple to
miss transformations that need to be made but aren't, resulting in
garbarge numbers.

(5) IN_UNIT(3600, "s/h")... well, I could see some use for that, but,
really, it's marginal, and a rule like "if the first argument is a
literal, then set unit metadata" introduces a lot of complexity, because
people might argue that "literal" includes something like
(3+0.5)*power(10, 3).  That, to me, seems rather unproportional for the
marginal gain.  So, for consistency and I'd say that should error out as
well.

If it's just at (5) that you disagree: Can you come up with some
language that distinuishes case (5) from the rest of them?

> I don't really like the idea of the function `FORCE_META`. Having a
> such function makes ADQL able to modify the formatted output: ADQL is

Well, the unit part of FORCE_META wouldn't be different from your
proposal for IN_UNIT, no?  I agree, though, that we should be careful
not to puncture abstraction too much.  Not having implemented
FORCE_META, I can't really say if that's already puncturing things too
much.

> convenient side-effect. In my mind the unit metadata setting ability
> of `IN_UNIT` is not its main goal. Its main goal is to convert a value
> from one unit to another, but ADQL has to say what to do if no unit is
> specified. Then, we have 2 choices: 1/ raise an error, 2/ set the unit
> to the given one. Raising an error seems a bit extrem for me and will
> probably disturb users since probably a lot of columns (from DB or
> UPLOAD), litterals or operations may not have unit. Then, I think it

The trouble is that you'll be hiding a lot of errors if you just trudge
on, and once it's happened, it'll be very difficult to trace back the
reason for odd values (if they are noticed at all -- as I said I expect
IN_UNIT will usually be used in largish cross-server query scripts).

Let's save our users grief by playing it safe and raising an error
whenever there are incompatible units, where "no unit" is incompatible
with all units (cf. NULL behaviour in SQL itself).

> is reasonnable to assume that the user SHOULD know what he is doing,
> scientifically speaking.

Now, if the user is a script querying lots of tables on lots of
servers, you can be sure that the user has no idea what it's doing...

**Comment by gmantele** [comment]
I think you misunderstood what I said until now (probably my explanations were
not clear enough, and if so, sorry): I never said that `IN_UNIT` should never
raise errors.

And, in all, I also think that we mixed up 3 things in this discussion:

1. the behaviour of `IN_UNIT` in function of the given value and unit
2. the result of `IN_UNIT`
3. how the unit (unitless included) is specified or not

-----

### 1. Behaviour of `IN_UNIT`

I completely agree with all your exemples from 1 to 4 (incl.). My only concern
has always been about (5).

So, to answer you:

> If it's just at (5) that you disagree: Can you come up with some
> language that distinuishes case (5) from the rest of them?

That was exactly the goal of this issue. Let me quote myself (in the
description of this GitHub issue):

> But what happens ...
> 
> * ... if no unit is set for this column reference?
> * ... if the value is not a column reference (e.g. a numeric litteral, a math
>   expression)?_

Nothing is stated in the current ADQL draft for these two cases. That's what I
wanted to fix.

My point has always been that an ADQL engine should not raise an error _IF_ no
unit is specified. If one is _set, somehow, as `unitless`_, then, an error
should be raised. I was never opposed to that latter, and I think it is indeed
the right thing to do. The current discussions are about the case 'no unit
set' (which is different from 'unitless' as we agreed).

For me, ADQL should just specify how `IN_UNIT` should behave in a generic way.
Based on your examples (which were exactly what I had in mind), I thought of
these 4 following cases:

|    Unit value     | IN_UNIT reaction |
|-------------------|------------------|
| correct VOUnit    | If target_unit is compatible with the unit of col1, perform the conversion. If not, raise an error. |
| INcorrect VOUnit  | Immediately raise an error. |
| unitless          | Immediately raise an error. |
| no unit specified | No error. Just apply target_unit on the output value (being exactly the given one). It is the responsibility of the user ; he has to know his science...ADQL can not do that for him. |

Before saying anything about this last case (corresponding to your (5)), let me just first precise my thoughts in the following part:

### 2. Result of `IN_UNIT`

Just tell me if you agree (or where you disagree) with the following
statement:

"""
The result of `IN_UNIT` is the given value converted into the given target
unit. The output value will have its unit metadata field set to the target unit
of `IN_UNIT`.

For instance: `IN_UNIT(foo, 'm/s')`, where foo in in 'km/s', will
be a value in 'm/s'.

If you do not do that you loose information. Such information _might_ be useful
for operations, or in the output query result (SELECT) for the user.
"""

_(of course, I would not write the last part like that in ADQL)_

------

That statement also lets me answer to you when you said:

> Well, the unit part of FORCE_META wouldn't be different from your
> proposal for IN_UNIT, no?  I agree, though, that we should be careful
> not to puncture abstraction too much.  Not having implemented
> FORCE_META, I can't really say if that's already puncturing things too
> much.

As you can see in my above statement, the ability of `IN_UNIT` to set the unit metadata field of a value is a _convenient side-effect_, as I said before. It is not what `IN_UNIT` is designed for in first place, in the contrary to your proposition with `FORCE_META`.

--------

### 3. "unitless" and "no unit set"

> Let's save our users grief by playing it safe and raising an error
> whenever there are incompatible units,

I completely agree with that.

> where "no unit" is incompatible
> with all units (cf. NULL behaviour in SQL itself).

...but not totaly here. In SQL, NULL is a special value that indicate the lack of value. That is perfect to say that a column has no unit specified, whereas an empty string is a valid way, according to VOUnits, to say that it is a "unitless" value.

However, nothing is stated about that in services or formats like TAP and VOTable. But according to me, is not the role of ADQL to specify that. So, I
think it is not reasonnable to continue further this discussion here.

This brings me to the 2 followings points:

* ADQL should not say anything about how to convert from one unit to another.
  That is the role of VOUnits, and so, maybe VOUnits could be updated in that
  direction.

* Besides, according to me, ADQL should not say how unitless or empty unit are
  specified. That is the role of standards like TAP and VOTable. Again, maybe
  these 2 standards can add a more precise note about unit specification.

Do you agree on these two points?

**Comment by gmantele** [comment]
Little report on the discussion we got, @msdemlei and me, at the Escape TechForum (from 4th to 6th Feb. 2020 in Strasbourg):

* `IN_UNIT` must raise an error when a unit is incorrect or when the
  source and target units are incompatible.

* It has to be noted that a unitless value can not be converted.
  Similarly a non-unitless value can not be converted to a unitless
  one. In such cases `IN_UNIT` must raise an error.

* A service providing the `IN_UNIT` function _should_ also support
  unit inference as much as possible.

* When not possible for the value to convert with `IN_UNIT`, two
  cases have to be distinguished:

  - a variable numeric expression : an error must be raised.
  - a constant numeric expression : `IN_UNIT` must return the same
    value as provided. A unit inference mechanism must then be able
    to set its unit metadata to the target unit.

  _**Note:** A_ variable _numeric expression is a numeric expression including a column reference. In opposition, a_ constant _numeric expression does not use any column reference ; it can be a literal, an operation, a function call or any combination of those._

**Comment by gmantele** [comment]
Finally, I realized, mostly by writing it with examples, that this new idea was probably too complex. Especially, the unitless aspect of so many expressions (mainly functions) prevents most of the possible uses of `IN_UNIT()`.

For instance: `IN_UNIT(IN_UNIT(4, 'g'), 'kg')`. In theory, it works, but in practice, it is an heavy writing (`4g` would be more appropriate, as suggested by @gervandiepen) but mainly raises an inconsistency: we say that unitless expressions are forbidden, but we allow just numeric literals. And what about functions? As literals, they have no way to convey the unit information by themselves.

So, I changed a bit my proposal. For functions, a unit inference mechanism must be used in order to resolve the unit of the result, otherwise an error must be thrown. For literals, since there is no way to specify their unit, they are, by facts, unitless, and so, an error must be thrown as well.

I wrote my updated proposal, with some examples, in the PR #27 . Could anyone interested by this subject have a look?
